crypto/ocsp/ocsp_print.c

```diff
@@ -62,3 +78,198 @@ const char *OCSP_crl_reason_str(long s) {
   size_t tbl_size = (sizeof(reason_tbl) / sizeof((reason_tbl)[0]));
   return do_table2string(s, reason_tbl, tbl_size);
 }
+
+int OCSP_REQUEST_print(BIO *bp, OCSP_REQUEST *req, unsigned long flags) {
+  if(bp == NULL|| req ==NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  long l;
+  OCSP_CERTID *cid = NULL;
+  OCSP_ONEREQ *one = NULL;
+  OCSP_REQINFO *inf = req->tbsRequest;
+  OCSP_SIGNATURE *sig = req->optionalSignature;
+
+  if (BIO_puts(bp, "OCSP Request Data:\n") <= 0) {
+    return 0;
+  }
+  l = ASN1_INTEGER_get(inf->version);
+  if (BIO_printf(bp, "    Version: %ld (0x%ld)", l + 1, l) <= 0) {
+    return 0;
+  }
+  if (inf->requestorName != NULL) {
+    if (BIO_puts(bp, "\n    Requestor Name: ") <= 0) {
+      return 0;
+    }
+    GENERAL_NAME_print(bp, inf->requestorName);
+  }
+  if (BIO_puts(bp, "\n    Requestor List:\n") <= 0) {
+    return 0;
+  }
+  for (size_t i = 0; i < sk_OCSP_ONEREQ_num(inf->requestList); i++) {
+    one = sk_OCSP_ONEREQ_value(inf->requestList, i);
+    cid = one->reqCert;
+    ocsp_certid_print(bp, cid, 8);
+    if (!X509V3_extensions_print(bp, "Request Single Extensions",
+                                 one->singleRequestExtensions, flags, 8)) {
+      return 0;
+    }
+  }
+  if (!X509V3_extensions_print(bp, "Request Extensions", inf->requestExtensions,
+                               flags, 4)) {
+    return 0;
+  }
+  if (sig != NULL) {
+    X509_signature_print(bp, sig->signatureAlgorithm, sig->signature);
+    for (size_t i = 0; i < sk_X509_num(sig->certs); i++) {
+      X509_print(bp, sk_X509_value(sig->certs, i));
+      PEM_write_bio_X509(bp, sk_X509_value(sig->certs, i));
+    }
+  }
+  return 1;
+}
+
+int OCSP_RESPONSE_print(BIO *bp, OCSP_RESPONSE *resp, unsigned long flags) {
+  if(bp == NULL|| resp ==NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
```

Comment:
This raises an interesting question: is the semantics of printing a null response printing the word "null response" to the bio (and returning success), or returning a null parameter error (and returning an error)?

Reply:
Good points, I'll leave it the way it is since it doesn't really make sense to pass nothing to print in the first place.
We're doing better than OpenSSL is here anyways, they would segfault if a NULL is passed in.